783 admin console configurator

abidingotter opened the pull request:
To run it:
`./gradlew :configurator:run`

Note that the config json it creates is stored in [user]/openequella-admin-console directory

PenghaiZhang commented on configurator/src/main/java/org/apereo/openequella/adminconsole/configurator/Configurator.java:
```diff
@@ -0,0 +1,288 @@
+            removeServerButton = new JButton("Remove Server");
+            editProxyButton = new JButton("Proxy Settings...");
+
+            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
```
We need to set LookAndFeel before instantiating above components.

PenghaiZhang commented on configurator/src/main/java/org/apereo/openequella/adminconsole/configurator/Configurator.java:
```diff
@@ -0,0 +1,288 @@
+            final int[] rows = { height1, buttonsPanel.getPreferredSize().height };
+            final int[] cols = { WINDOW_WIDTH };
+
+            final TableLayout layout = new TableLayout(rows, cols);
```
As this UI should not be very complicated, can we consider using other  layout such as GridLayout? I personally feel TableLayout may not be the best choice because of more configurations required.

abidingotter replied:
I think TableLayout is pretty simple to use.  I don't think it's worthwhile wrestling with GridLayout to make it look ok.

PenghaiZhang commented on common/src/main/java/org/apereo/openequella/adminconsole/config/ServerProfile.java:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	public void setUuid(String uuid) {
+		this.uuid = uuid;
```
Because uuid is randomly generated in constructor so maybe setUuid is not needed?

abidingotter replied:
Also pretty sure this one is needed by the Json deserialiser.

abidingotter replied:
Turns out it is not needed, so I will remove it :)

PenghaiZhang commented on configurator/src/main/java/org/apereo/openequella/adminconsole/configurator/Configurator.java:
```diff
@@ -0,0 +1,288 @@
+        if (index > -1) {
+            final ServerProfile profile = (ServerProfile) serverPicker.getSelectedItem();
+            final int result = JOptionPane.showConfirmDialog(this, "Are you sure you want to remove this server?",
+                    "Are you sure?", JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
```
Just remind that these strings need to be declared as constants :)

abidingotter replied:
Not sure it was a big deal, but I have done so :)

PenghaiZhang commented on configurator/src/main/java/org/apereo/openequella/adminconsole/configurator/ProxySettingsDialog.java:
```diff
@@ -0,0 +1,233 @@
+    proxySettings.setPort(portModel.getNumber().intValue());
+
+    if (credentials.isSelected()) {
+      proxySettings.setUsername(usernameField.getText());
```
I think it might be better to check if usernameField.getText().trim().length() > 0 and then call setUsername

abidingotter replied:
I guess when you do it that way the user might think that they *need* to enter a username and password.

edalex-ian commented:
At a basic level, this fails to do a standard `gradlew build` - which in turn will fail the Travis CI build when it get's merged:

![2019-04-18-172449_1515x625_scrot](https://user-images.githubusercontent.com/43919233/56343819-22358b80-61ff-11e9-81bd-3846d7de21b7.png)

I also have a copy of queries around the code. So please wait my review before merging.

abidingotter commented on common/src/main/java/org/apereo/openequella/adminconsole/config/ProxySettings.java:
```diff
@@ -0,0 +1,83 @@
+  private String password;
+  private String obsPassword;
+
+  public ProxySettings() {
```
It's needed by the Json deserialiser.  At least I am pretty sure it is.  I can remove it and see if it still works.

abidingotter commented:
> At a basic level, this fails to do a standard gradlew build - which in turn will fail the Travis CI build when it get's merged:
> 
> Also note the strong warning that this wont work with Gradle 6, so we better address those issues now.
> I also have a copy of queries around the code. So please wait my review before merging.

Gradle build is fixed, however I have no idea on how to get rid of the warnings (other than removing the linter plugin I added ;) )

abidingotter commented:
> Discuss this approach, we probably need a better method for the code sharing;
> Address the gradle deprecation issues - and get ./gradlew build successful;
> Possibly rebase on develop so that it's got the Travis CI in it;
> Look at removing Guava as it seems overkill just for a couple of @Nullable annotations; and
> Add the Apereo headers to all the source files - arguably we should add a gradle plugin for this (perhaps https://github.com/hierynomus/license-gradle-plugin).


1. Not convinced we need to do anything about the code sharing.  Including the launcher code that's coming, it's a probably a dozen files (that haven't changed in the main repo for years).
The only solutions I could see are:
Creating a git repo for the common code and using git submodules *shudder*. We moved away from submodules for a reason.
Putting the common code into a jar on a private Maven repo, or even Maven Central.  A pretty massive undertaking for little gain.
Making the build script copy these files in from the main repo.  Eek.

2. Fixed gradle build, but not sure what to do about the warnings.

3. Travis build is included now

4. Have removed Guava

5. Will look at the license plugin

edalex-ian commented:
>  1. Not convinced we need to do anything about the code sharing.  Including the launcher code that's coming, it's a probably a dozen files (that haven't changed in the main repo for years).

But as we know, copy-paste is a great way to create some maintenance nightmares - so if anything the fact it's several files makes it more concerning than one block of text in multiple places. But perhaps as a starting point to get this done we can live with it if all the unused code (which thankfully IntelliJ can identify very quickly) is removed. Maybe it's also mitigated by the hope that in the not too distant future we'll have all the admin features in Web UI.

>  The only solutions I could see are:
> 
>  * Creating a git repo for the common code and using git submodules _shudder_. We moved away from submodules for a reason.

Agree, we wouldn't want to go there.

>  * Putting the common code into a jar on a private Maven repo, or even Maven Central.  A pretty massive undertaking for little gain.

This could be an option, and is not too difficult other than what repo to use being an open source project.
 
>  * Making the build script copy these files in from the main repo.  Eek.

Yeah, let's not go there. :)

**So what next?**

If you @abidingotter can just go and remove all the cruft then we should be able to merge away. As for the header stuff, I'll ask @PenghaiZhang to work on that.

edalex-ian commented:
Actually, on the code sharing front, there is a way to do it from git repos with tags and Gradle (and even SBT if we had to do it on the oEQ side):

https://jitpack.io/

But that's just an FYI. Let's let it be for now.

abidingotter commented:
> But as we know, copy-paste is a great way to create some maintenance nightmares - so if anything the fact it's several files makes it more concerning than one block of text in multiple places. But perhaps as a starting point to get this done we can live with it if all the unused code (which thankfully IntelliJ can identify very quickly) is removed. Maybe it's also mitigated by the hope that in the not too distant future we'll have all the admin features in Web UI.

Agreed, I don't condone copy and paste in general, but since it was a small stop-gap app with a dozen copied classes I felt it was a special case.  The alternative was to spend a few days getting some sort of code sharing working. 

> If you @abidingotter can just go and remove all the cruft then we should be able to merge away (looks like you did remove a good number of unused classes and files in last couple of commits - so maybe this is mostly done). As for the header stuff, I'll ask @PenghaiZhang or @SammyIsConfused to work on that - we'll do in another branch.

There shouldn't be any unused code now.  I'll perform the merge.

abidingotter commented:
> Actually, on the code sharing front, there is a way to do it from git repos with tags and Gradle (and even SBT if we had to do it on the oEQ side):
> https://jitpack.io/
> But that's just an FYI. Let's let it be for now.
>  jitpackJitPack | Publish JVM and Android librariesJitPack makes it easy to release your Java or Android library. Publish straight from GitHub or Bitbucket.

Jitpack looks cool, but it would require refactoring our Equella code to make a common, gradle-built library.  Not a small undertaking...
